quantile/tdigest/tdigest.go

```diff
@@ -0,0 +1,285 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tdigest
+
+import (
+	"math"
+	"sort"
+)
+
+var (
+	positiveInfinity = math.Inf(1)
+	negativeInfinity = math.Inf(-1)
+	sentinelCentroid = Centroid{Mean: positiveInfinity, Weight: 0.0}
+)
+
+type centroidsByMeanAsc []Centroid
+
+func (c centroidsByMeanAsc) Len() int           { return len(c) }
+func (c centroidsByMeanAsc) Less(i, j int) bool { return c[i].Mean < c[j].Mean }
+func (c centroidsByMeanAsc) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+
+type tDigest struct {
+	compression      float64       // compression factor
+	mergedCapacity   int           // merged centroid slice capacity
+	unmergedCapacity int           // unmerged centroid slice capacity
+	multiplier       int64         // quantile precision multiplier
+	centroidsPool    CentroidsPool // centroids pool
+
+	closed         bool       // whether the t-digest is closed
+	merged         []Centroid // merged centroids
+	mergedWeight   float64    // total weight of merged centroids
+	unmerged       []Centroid // unmerged centroid slice capacity
+	unmergedWeight float64    // total weight of unmerged centroids
+	minValue       float64    // minimum value
+	maxValue       float64    // maximum value
+}
+
+// mergedCapacity computes the capacity of the merged centroid slice
+func mergedCapacity(compression float64) int {
+	return int(math.Pi*compression/2 + 0.5)
+}
+
+func unmergedCapacity(compression float64) int {
+	compression = math.Min(math.Max(20, compression), 1000)
+	return (int)(7.5 + 0.37*compression - 2e-4*compression*compression)
+}
+
+// NewTDigest creates a new t-digest
+// TODO(xichen): add pooling for t-digests
+// TODO(xichen): add metrics
+func NewTDigest(opts Options) TDigest {
+	centroidsPool := opts.CentroidsPool()
+	compression := opts.CompressionFactor()
+	mergedCapacity := mergedCapacity(compression)
+	unmergedCapacity := unmergedCapacity(compression)
+
+	var multiplier int64
+	if precision := opts.QuantilePrecision(); precision != 0 {
+		multiplier = int64(math.Pow10(precision))
+	}
+
+	return &tDigest{
+		compression:      opts.CompressionFactor(),
+		multiplier:       multiplier,
+		centroidsPool:    centroidsPool,
+		merged:           centroidsPool.Get(mergedCapacity),
+		mergedCapacity:   mergedCapacity,
+		unmerged:         centroidsPool.Get(unmergedCapacity),
+		unmergedCapacity: unmergedCapacity,
+		minValue:         positiveInfinity,
+		maxValue:         negativeInfinity,
+	}
+}
+
+func (d *tDigest) Merged() []Centroid {
+	return d.merged
+}
+
+func (d *tDigest) Unmerged() []Centroid {
+	return d.unmerged
+}
+
+func (d *tDigest) Add(value float64) {
+	d.add(value, 1.0)
+}
+
+func (d *tDigest) Quantile(q float64) float64 {
+	// compress the centroids first
+	d.compress()
+
+	var (
+		targetWeight = q * d.mergedWeight
+		curWeight    = 0.0
```

Comment:
i.e. some places we use `curThing` and other places we use `currThing` perhaps standardize on one?